soh/soh/Enhancements/gameplaystats.cpp

```diff
@@ -174,6 +179,151 @@ void DisplayTimeHHMMSS(uint32_t timeInTenthsOfSeconds, std::string text, ImVec4
+    SaveManager::Instance->LoadData("heartContainers", gSaveContext.sohStats.heartContainers);
+    SaveManager::Instance->LoadArray("dungeonKeys", ARRAY_COUNT(gSaveContext.sohStats.dungeonKeys), [](size_t i) {
+        SaveManager::Instance->LoadData("", gSaveContext.sohStats.dungeonKeys[i]);
+    });
+    SaveManager::Instance->LoadData("playTimer", gSaveContext.sohStats.playTimer);
+    SaveManager::Instance->LoadData("pauseTimer", gSaveContext.sohStats.pauseTimer);
+    SaveManager::Instance->LoadArray(
+        "itemTimestamps", ARRAY_COUNT(gSaveContext.sohStats.itemTimestamp),
+        [](size_t i) { SaveManager::Instance->LoadData("", gSaveContext.sohStats.itemTimestamp[i]); });
+    SaveManager::Instance->LoadArray(
+        "sceneTimestamps", ARRAY_COUNT(gSaveContext.sohStats.sceneTimestamps), [](size_t i) {
+            SaveManager::Instance->LoadStruct("", [&i]() {
+                int scene, room, sceneTime, roomTime, isRoom;
+                SaveManager::Instance->LoadData("scene", scene);
+                SaveManager::Instance->LoadData("room", room);
+                SaveManager::Instance->LoadData("sceneTime", sceneTime);
+                SaveManager::Instance->LoadData("roomTime", roomTime);
+                SaveManager::Instance->LoadData("isRoom", isRoom);
+                if (scene == 0 && room == 0 && sceneTime == 0 && roomTime == 0 && isRoom == 0) {
+                    return;
+                }
+                gSaveContext.sohStats.sceneTimestamps[i].scene = scene;
+                gSaveContext.sohStats.sceneTimestamps[i].room = room;
+                gSaveContext.sohStats.sceneTimestamps[i].sceneTime = sceneTime;
+                gSaveContext.sohStats.sceneTimestamps[i].roomTime = roomTime;
+                gSaveContext.sohStats.sceneTimestamps[i].isRoom = isRoom;
+            });
+        });
+    SaveManager::Instance->LoadData("tsIdx", gSaveContext.sohStats.tsIdx);
+    SaveManager::Instance->LoadArray("counts", ARRAY_COUNT(gSaveContext.sohStats.count), [](size_t i) {
+        SaveManager::Instance->LoadData("", gSaveContext.sohStats.count[i]);
+    });
+    SaveManager::Instance->LoadArray(
+        "scenesDiscovered", ARRAY_COUNT(gSaveContext.sohStats.scenesDiscovered),
+        [](size_t i) { SaveManager::Instance->LoadData("", gSaveContext.sohStats.scenesDiscovered[i]); });
+    SaveManager::Instance->LoadArray(
+        "entrancesDiscovered", ARRAY_COUNT(gSaveContext.sohStats.entrancesDiscovered),
+        [](size_t i) { SaveManager::Instance->LoadData("", gSaveContext.sohStats.entrancesDiscovered[i]); });
+    SaveManager::Instance->LoadArray(
+        "locationsSkipped", ARRAY_COUNT(gSaveContext.sohStats.locationsSkipped),
+        [](size_t i) { SaveManager::Instance->LoadData("", gSaveContext.sohStats.locationsSkipped[i]); });
+}
+
+void SaveStats(SaveContext* saveContext, const std::string& subSection) {
+    if (subSection == "all") {
```

Comment:
Can't do an enum because it has to be extensible for mods later on.

Reply:
Whatever system ends up being used for the cvar rework would probably work here, too, but until then, I think this is the best we can get. I'm having trouble thinking of anything else we could do.

Reply:
As long as it's easy to swap out where you have strings later, it's honestly not a big deal to do it with strings for now given how many other opportunities we have for performance improvement because of them :)  In fact, if mods will have access to this, that almost makes it MORE from just an error-prevention standpoint to figure out something better than strings before mods start using it.  Maybe start a Discord thread to noodle on this specifically?

Reply:
I actually was getting an idea, but I'm having trouble with some of the logistics of it. Basically, when `AddSaveFunc` is called, what it would do is check against a database of other section names already registered, and if unique, add the function information to a data structure and return an index that can be used in future calls to `SaveSection`. The name registry would make sense to have as a vector, but I can't quite figure out how to structure the rest of it, as maps don't have inherent indexes, but arrays need to be statically sized. But I would want that structure to store the section name, the function reference, and the game save tie-in.

Reply:
And we could have an enum of built-in sections (like sohStats, trackerData) and relevant subsections, that each save function is pre-registered with, and start the index for the mod functions from the max enum value. Since the sections will still be stored by name in the save file, what index they end up getting (due to differences in mod load order) shouldn't matter.

Reply:
Why does the array need to be statically sized?  Or I guess if it does, you can easily wrap it in a span since they can automatically detect the length of the underlying array.  That's what I did in the Imgui cleanup to get rid of all the array length arguments that the comboboxes had:  https://github.com/jbodner09/Shipwright/blob/49d89d2759aad61d2be75401d834d5587777f8b9/soh/soh/UIWidgets.cpp#LL226C1-L250C10  And then any of the related info like the name string, save function, etc., you could just stuff into a little private class to store instances of in that list.  

So, if I'm following you correctly, you just have a vector of all different save sections, where each entry in the vector is your custom class.  On `AddSaveFunc`, it scans the vector and checks each entry's string, and if it finds one that matches, it returns the index of that item so the caller just needs to call `saveList[index].saveSection` as necessary later on, and if it doesn't find one, it just adds a new entry to the end of the vector and returns the index of that new item.

Reply:
Well, that's the thing. If it's not an array, then there's no inherent ID associated with the data structure that contains the rest of the saveSection info. But you can't dynamically size arrays; when declaring one, it requires a size parameter. So it's the question of tying that section name to an ID that could then be passed back and forth as a parameter. The structure I came up with originally that I'm working off of for now is a `map<string, int>` for name to ID mapping, then another one of `<int, SaveFuncInfo>` that ties ID to the struct, where `SaveFuncInfo` has name, version, function, and game save tie-in values.

Reply:
When `AddSaveFunc` is called, then, it checks to see if the name is already in the name->ID map. If it is, it returns an error (it shouldn't automatically return the ID associated with it because it might not be the SaveFunc they're looking for). If not, it grabs the next available integer for ID, adds the name->ID combo to that first map, then constructs the `SaveFuncInfo` object to insert into the second map. The nice thing about this system is that you can register subsections with the same `SaveFunc`, store those subsection IDs locally, then check the section ID passed through the `SaveFunc` hook to see which data the function should save. And the name can then be whatever you want. With `entrancesDiscovered` for instances, I could register the subsection as "sohStats.entrances" or "entrancesDiscovered".

Reply:
Should it be safe to assume that the "base" section being passed as the sectionID to SaveFile means entire game save?

Reply:
@Malkierian yeah, defaulting to everything seems like a good fallback.  I guess I still don't see what the issue with using a `std::vector` is.  Vectors are dynamically sized, so you can always add onto the end of them, and they can be accessed via indices just like arrays or using the at() function.  But if it's working with maps, I guess it's not important!